eth_account/_utils/signing.py

```diff
@@ -44,7 +58,182 @@ def sign_transaction_dict(eth_key, transaction_dict):
+    where each member is written as type ‖ " " ‖ name.
+    """
+    # Getting the dependencies and sorting them alphabetically as per EIP712
+    deps = dependencies(primaryType, types)
+    deps_without_primary_type = list(filter(lambda x: x != primaryType, deps))
+    sorted_deps = [primaryType] + sorted(deps_without_primary_type)
+
+    result = ''.join(
+        [
+            struct_name + "(" + ','.join(map(dict_to_type_name_converter, types[struct_name])) + ")"
+            for struct_name in sorted_deps
+        ]
+    )
+    return result
+
+
+def typeHash(primaryType, types):
+    return keccak(text=encodeType(primaryType, types))
+
+
+def encodeData(primaryType, types, data):
+    encTypes = []
+    encValues = []
+
+    # Add typehash
+    encTypes.append("bytes32")
+    encValues.append(typeHash(primaryType, types))
+
+    # Add field contents
+    for field in types[primaryType]:
+        value = data[field["name"]]
+        if field["type"] == "string":
+            if not isinstance(value, str):
+                raise TypeError(
+                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
+                    "string value".format(
+                        field["name"],
+                        primaryType,
+                        value,
+                        type(value),
+                    )
+                )
+            # Special case where the values need to be keccak hashed before they are encoded
+            encTypes.append("bytes32")
+            hashed_value = keccak(text=value)
+            encValues.append(hashed_value)
+        elif field["type"] == "bytes":
+            if not isinstance(value, bytes):
+                raise TypeError(
+                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
+                    "bytes value".format(
+                        field["name"],
+                        primaryType,
+                        value,
+                        type(value),
+                    )
+                )
+            # Special case where the values need to be keccak hashed before they are encoded
+            encTypes.append("bytes32")
+            hashed_value = keccak(primitive=value)
+            encValues.append(hashed_value)
+        elif field["type"] in types:
+            # This means that this type is a user defined type
+            encTypes.append("bytes32")
+            hashed_value = keccak(primitive=encodeData(field["type"], types, value))
+            encValues.append(hashed_value)
+        elif field["type"][-1] == "]":
+            # TODO: Replace the above conditionality with Regex for identifying arrays declaration
+            raise NotImplementedError("TODO: Arrays currently unimplemented in encodeData")
+        else:
+            # First checking to see if the individual values can be encoded
+            try:
+                is_encodable(field["type"], value)
+            except:
```

Comment:
Does not appear that `is_encodable` raises exceptions.

https://github.com/ethereum/eth-abi/blob/6aa2bd0e3a5f213afd4d475dc6f0689b31875637/eth_abi/codec.py#L89-L114

Reply:
@pipermerriam The exception is raised here 
https://github.com/ethereum/eth-abi/blob/6aa2bd0e3a5f213afd4d475dc6f0689b31875637/eth_abi/codec.py#L102.

This is triggered when we give something like `is_encodable("hello", "something")`

Reply:
Ah!

I opened https://github.com/ethereum/eth-abi/issues/125to track the upstream issue

For now how about we add an `is_valid_type` utility that **always** returns `True/False`, underlying implementation can depend on the exception raising behavior from `is_encodable`.  Make sure to link to the `eth-abi` issue in that utility function's docstring so that it's clear why it exists and also so we can know when to remove it once that issue has been addressed in `eth-abi`.

Reply:
cc @davesque in case you hadn't seen this interesting application of a custom eth-abi registry: the typed signed message EIP 712.

Reply:
Skimmed over it just now.  Nifty.  Looking forward to seeing how the registry gets used for that.